nix/web-security-tracker.nix

```diff
@@ -113,42 +113,81 @@ in {
     };
 
-    systemd.services.web-security-tracker-server = {
-      description = "A web security tracker ASGI server";
-      after = [ "network.target" "postgresql.service" ];
-      requires = [ "postgresql.service" ];
-      wantedBy = [ "multi-user.target" ];
-      path = [ pythonEnv wstManageScript ];
-      serviceConfig = {
-        User = "web-security-tracker";
-        Restart = "always";
-        WorkingDirectory = "/var/lib/web-security-tracker";
-        StateDirectory = "web-security-tracker";
-        RuntimeDirectory = "web-security-tracker";
-        LoadCredential = credentials;
+    systemd.services = {
+      web-security-tracker-server = {
+        description = "A web security tracker ASGI server";
+        after = [ "network.target" "postgresql.service" ];
+        requires = [ "postgresql.service" ];
+        wantedBy = [ "multi-user.target" ];
+        path = [ pythonEnv wstManageScript ];
+        serviceConfig = {
+          User = "web-security-tracker";
+          Restart = "always";
+          WorkingDirectory = "/var/lib/web-security-tracker";
+          StateDirectory = "web-security-tracker";
+          RuntimeDirectory = "web-security-tracker";
+          LoadCredential = credentials;
+        };
+        environment = {
+          DATABASE_URL = databaseUrl;
+          USER_SETTINGS_FILE = "${configFile}";
+        };
+        preStart = ''
+          # Auto-migrate on first run or if the package has changed
+          versionFile="/var/lib/web-security-tracker/package-version"
+          if [[ $(cat "$versionFile" 2>/dev/null) != ${cfg.package} ]]; then
+            wst-manage migrate --no-input
+            wst-manage collectstatic --no-input --clear
+            echo ${cfg.package} > "$versionFile"
+          fi
+        '';
+        script = let
+          networking = if cfg.unixSocket != null then
+            "-u ${cfg.unixSocket}"
+          else
+            "-b 127.0.0.1 -p ${toString cfg.port}";
+        in ''
+          daphne ${networking} \
+            tracker.asgi:application
+        '';
       };
-      environment = {
-        DATABASE_URL = databaseUrl;
-        USER_SETTINGS_FILE = "${configFile}";
+
+      web-security-tracker-delta = {
+        description = "A web security tracker ASGI server";
+        after = [ "network.target" "postgresql.service" ];
+        requires = [ "postgresql.service" ];
+        wantedBy = [ "multi-user.target" ];
+        path = [ pythonEnv wstManageScript ];
+        serviceConfig = {
+          User = "web-security-tracker";
+          Restart = "always";
+          WorkingDirectory = "/var/lib/web-security-tracker";
+          StateDirectory = "web-security-tracker";
+          RuntimeDirectory = "web-security-tracker";
+          LoadCredential = credentials;
+          Type = "oneshot";
+        };
+        environment = {
+          DATABASE_URL = databaseUrl;
+          USER_SETTINGS_FILE = "${configFile}";
+        };
+        preStart = ''
```

Comment:
Order it after the server script IMHO otherwise we may run into concurrent migrations…

Reply:
(or before the load, however you want)